### DISCUSSION TITLE
Linearization files

### DISCUSSION THREAD

**Discussion by joeb8799** [NatLabRockies/ROSCO#146] [General]
Hello,
I need to know how can i get the linearization files (.lin) used as input in example 12.
I need to get .lin files for a turbine model that i have so that i can apply the method of example 12.
Is there any guide on how to use OpenFAST in order to get the .lin files

**Comment by dzalkind** [comment]
Unfortunately, there aren't guides, but this page might be a good start: https://github.com/OpenFAST/openfast/issues/738

You can also begin by setting the `Linearize` flag to `True` in the main OpenFAST input file.  The other options in that section have good descriptions, as well.

**Reply by joeb8799** [reply]
Hello,
Thanks for your answer. I now know how to compile openfast in order to get the linearization files, but what I need to know is the following:
what should I change in the input files (.fst, servodyn, aerodyn ...) so that I can get the same .fst files used in example 12, so that I can use them to apply the robust scheduling for my turbine model.
Can anyone help me or can I talk to the person who generated these .lin files?

**Reply by nikhar-abbas** [reply]
Hi,

I generated those .lin files a while back and, unfortunately, do not have the _exact_ OpenFAST input deck that I used to generate them. Here is a set of very similar input files should work with OpenFAST v3.0.0:
[iea15mw_baseline.tar.gz](https://github.com/NREL/ROSCO/files/8942525/iea15mw_baseline.tar.gz)

Hopefully this gets you started. You should be able to modify the input deck for different degrees of freedom and wind speeds easily from here.

- Nikhar

**Comment by joeb8799** [comment]
Thank you very much, that was very helpful, I managed to get the linearization files for wind speed between 12 and 24 m/s but I have this problem for speeds between 4 and 11:
OpenFAST

Copyright (C) 2022 National Renewable Energy Laboratory
Copyright (C) 2022 Envision Energy USA LTD

This program is licensed under Apache License Version 2.0 and comes with ABSOLUTELY NO WARRANTY.
See the "LICENSE" file distributed with this software for details.
**************************************************************************************************

OpenFAST-v3.1.0
Compile Info:
 - Compiler: Intel(R) Fortran Compiler 1900
 - Architecture: 64 bit
 - Precision: single
 - OpenMP: No
 - Date: Mar 08 2022
 - Time: 10:53:02
Execution Info:
 - Date: 06/21/2022
 - Time: 20:25:36+0300

OpenFAST input file heading:
    Generated with AeroElasticSE FAST driver

Running ElastoDyn.
Nodal outputs section of ElastoDyn input file not found or improperly formatted.
Running AeroDyn.
AD15 Nodal Outputs: Nodal output section of AeroDyn input file not found or improperly formatted.
Skipping nodal outputs.
Running InflowWind.
Running HydroDyn.
  Setting WaveTMax to 0.0 since WaveMod = 0
 Reading in WAMIT output with root name ".\HydroData/IEA-15-240-RWT-UMaineSemi".
 Computing radiation impulse response functions and wave diffraction forces.

Using SS_Excitation Module, with 36 excitation states

Using SS_Radiation Module, with 44 of 44 radiation states

MAP++ environment properties (set externally)...
    Gravity constant          [m/s^2]  : 9.81
    Sea density               [kg/m^3] : 1025.00
    Water depth               [m]      : 200.00
    Vessel reference position [m]      : 0.00 , 0.00 , 0.00

Running ServoDyn.
 Time: 290 of 5000 seconds.  Estimated final completion at 20:55:38.
The BEM solution is being turned off due to low TSR.  (TSR = 1.9987). This warning will not be
 Time: 310 of 5000 seconds.  Estimated final completion at 20:55:52.

FAST_Solution:FAST_AdvanceStates:SrvD_UpdateStates:CalculateTorque:u%HSS_Spd is negative. Simple
variable-speed control model is not valid for motoring situations.

 OpenFAST encountered an error at simulation time 316.87 of 5000 seconds.
 Simulation error level: FATAL ERROR

 Aborting OpenFAST.
Do u have any idea how to fix it? 
Sorry for bothering you again.
Have a good day.

**Reply by nikhar-abbas** [reply]
It is hard to say exactly what the problem is without looking at the time histories of the failing simulations.

Two things:
1) the `TrimGain` parameter in the .fst fail may need to be changed so that the blades pitch reasonably. It looks like this may be wrong and causing the turbine to spin backwards.
2) You will need to ensure that the parameters in the `SIMPLE VARIABLE-SPEED TORQUE CONTROL` section of the ServoDyn file are properly defined. You can pull some of these values from the DISCON.IN file that the ROSCO tuning processes define.

Cheers,
Nikhar

**Comment by joeb8799** [comment]
Hello Mr Abbas,
Sorry to bother you again. I am trying to get the linearization files for the IEA-15-240-RWT-UMaineSemi turbine model that we have in the ROSCO test_cases. I tried to modify the inputs in a similar manner to the example you sent me last week but got the error below. Can you help me fix it ?
OpenFAST

Copyright (C) 2022 National Renewable Energy Laboratory
Copyright (C) 2022 Envision Energy USA LTD

This program is licensed under Apache License Version 2.0 and comes with ABSOLUTELY NO WARRANTY.
See the "LICENSE" file distributed with this software for details.
**************************************************************************************************

OpenFAST-v3.1.0
Compile Info:
 - Compiler: Intel(R) Fortran Compiler 1900
 - Architecture: 64 bit
 - Precision: single
 - OpenMP: No
 - Date: Mar 08 2022
 - Time: 10:53:02
Execution Info:
 - Date: 06/27/2022
 - Time: 09:51:07+0300

OpenFAST input file heading:
    IEA 15 MW offshore reference model on UMaine VolturnUS-S semi-submersible floating platform

Running ElastoDyn.
Nodal outputs section of ElastoDyn input file not found or improperly formatted.
Running AeroDyn.
AD15 Nodal Outputs: Nodal output section of AeroDyn input file not found or improperly formatted.
Skipping nodal outputs.
Running InflowWind.
Running ServoDyn.

FAST_InitializeAll:InitModuleMappings:ED_2_AD_BladeMotion(1):MeshMapCreate:CreateMotionMap_L2_to_L
2:CreateMapping_ProjectToLine2:Found close value for node 50. (1.99464E-05 m)

 Time: 0 of 5000 seconds.
Warning: SkewedWakeCorrection encountered a large value of chi (90.204 deg), so the yaw
correction will be limited. This warning will not be repeated though the condition may persist.
See the AD15 chi output channels, and consider turning off the Pitt/Peters skew model (set
SkewMod=1) if this condition persists.


FAST_Solution:FAST_AdvanceStates:AD_UpdateStates:BEMT_UpdateStates:UpdatePhi(node 8, blade
2):BEMT_UnCoupledSolve:There is no valid value of phi for these operating conditions: Vx =
2.1998, Vy = -21.278, rlocal = 20.515, theta = 0.19031, geometric phi = 3.0386. This warning will
not be repeated though the condition may persist. (See GeomPhi output channel.)

The BEM solution is being turned off due to low TSR.  (TSR = -972.62). This warning will not be
repeated though the condition may persist. (See GeomPhi output channel.)

FAST_Solution:FAST_AdvanceStates:ED_ABM4:ED_CalcContStateDeriv:SetCoordSy:Small angle assumption
violated in SUBROUTINE SmllRotTrans() due to a large blade deflection (ElastoDyn SetCoordSy). The
solution may be inaccurate. Simulation continuing, but future warnings from SmllRotTrans() will
be suppressed.
 Additional debugging message from SUBROUTINE SmllRotTrans(): 7.09 s
 Time: 10 of 5000 seconds.  Estimated final completion at 10:25:29.

FAST_Solution:FAST_AdvanceStates:AD_UpdateStates:SetDisturbedInflow:TwrInfl:getLocalTowerProps:Tow
er strike.

 OpenFAST encountered an error at simulation time 15.03 of 5000 seconds.
 Simulation error level: FATAL ERROR

 Aborting OpenFAST.
